### PR TITLE
Also try to trigger merge on approved PR review

### PR DESCRIPTION
**Pull request by jimmidyson**

### CODE REVIEW

**Comment by jimmidyson** [comment]
@rhuss You were interested in this bot - care to review? I know it's Go but it's a good way to learn!

**Comment by rhuss** [comment]
will try my best ;-)

**Comment by jimmidyson** [comment]
Thanks!

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (rhuss) [inline pkg/webhook/merge_on_review_approval_and_status_checks.go:44]: guess I have to re-read how switch and casting works in go ;-) Maybe `payload` better named `eventObject` or so to better indicate the purpose ?

Comment (rhuss) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:44]: What's the difference between a PullRequestEvent and a PullRequestReviewEvent (I mean isn't a PRRE also a PRE ? )

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (rhuss) [inline pkg/webhook/merge_on_review_approval_and_status_checks.go:59]: Why don't you check here also for the value of the label (but only for that the type is 'labeled' ?)

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (rhuss) [inline pkg/webhook/merge_on_review_approval_and_status_checks.go:50]: If you handle the approval here directly and somewhere else to add the label which then in turn triggers the other handler watching on new labe;s isn't it then merged twice ? guess not, but I dont understand it really probably ;-)

**Review by rhuss — COMMENTED** [review]
Tried my best and had some probably silly question ;-)

need to practice my go a bit more, just about looking up the switch statement again (knew it, forgot it ;-)

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:44]: Just different events... vs https://developer.github.com/v3/activity/events/types/#pullrequestevent https://developer.github.com/v3/activity/events/types/#pullrequestreviewevent

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:50]: It's safe to try to merge something twice, only one will succeed. Guess I could check merge status of PR in this case, but due to async nature of webhooks not sure that will protect against multiple attempts to merge PR, bit of a race condition.

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:59]: The event doesn't contain what label has been added and you have to get the issue (a PR is also an issue...) which contains the list of labels... this is done in the called merge function so no need to repeat that here.

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:44]: Renamed to `eventObject` - much better name.

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (rhuss) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:50]: So if I understand you right you want to merge in case of two use case:

* Someone labels directly the PR as "approved" without doing an explicit approcal step.
* Someone approves vie "Approved" step in the review workflow.

I wonder whether the flow here to check the approval directly is necessary as an approval already triggers the generation of the label which in turn triggers the merge (case 1). Is it because its more direct (because it contains only 1 step instead of 2) ? Just an other safeguard, or ?

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (rhuss) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:59]: Just to be sure that I got it right:

* In the `mergePR` function there is no check a label value 'approved' but if every of the required status is set to 'success'.
* So setting a label manually differently than "approved" just don't succeed if the status of the review has been already set to "approved".
* Setting a label manually to "approved" ... hmm .. that doesn't really do anything extra like approving the review. 

This means that the use case above with directly labelling with "approved' is not really supported.

So this check is only to trigger a potential merge. Think I got it :) (sorry for thinking loud)

**Review by rhuss — DISMISSED** [review]
Looks good to me (albeit a bit redundant since I guess that a merge is always triggered twice now: Once because of the label added, once for the direct approval event)

**Comment by rhuss** [comment]
Ha, triggered now the use case for which this PR is: If someone does a manual labelling with "approve", this triggers a prMerge() but this doesn't succeed because because not all statuses are cleared. Then I do the approval step, but no label is added because it already exists. so no event, no merge.

Maybe one could make it a bit more straight forward: Just trigger only on a PullRequestReviewEvent and then do the merge + labelling (for documentation purposes).

**Comment by rhuss** [comment]
Now for another scenarion: You approve but not all required status checks are happy (other than the one for the approval).

So I guess one would need yet another handler listening for status changes and do a merge always if all status checks succeed ?

**Comment by rhuss** [comment]
Sorry, haven't seen `autoMerger.handleStatusEvent()`. Forgot my last comment.

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:50]: It's just a safeguard. On a previous issue, I added `approved` label manually and then approved PR and no automerge was kicked off so it's belt and braces approach :)

File: pkg/webhook/merge_on_review_approval_and_status_checks.go
Comment (jimmidyson) [reply pkg/webhook/merge_on_review_approval_and_status_checks.go:59]: There should be a check for the `approved` label too... Ideally it would check that all PR reviews pass but GitHub integrations can't access reviews API endpoints (full list of accessible endoints [here](https://developer.github.com/early-access/integrations/available-endpoints/)).

**Review by rhuss — APPROVED** [review]
Looks even better now ;-)

**Comment by rhuss** [comment]
Hmm. old PuRe bot is still confused.

Merging it manually now

**Comment by jimmidyson** [comment]
Old PuRe bot was confused because your review went in while status checks were still running so didn't merge. Should have merged on status check passing  - but that requires the fix in #2 to be deployed which it isn't yet.

Thanks for your reviews @rhuss!

**Pull request merged**
